source/extensions/filters/network/redis_proxy/conn_pool_impl.cc

```diff
@@ -122,6 +138,91 @@ InstanceImpl::ThreadLocalPool::makeRequest(const std::string& key,
+
+  return client->redis_client_->makeRequest(request, callbacks);
+}
+
+Common::Redis::Client::PoolRequest*
+InstanceImpl::ThreadLocalPool::makeRequestToHost(const std::string& host_address,
+                                                 const Common::Redis::RespValue& request,
+                                                 Common::Redis::Client::PoolCallbacks& callbacks) {
+  if (cluster_ == nullptr) {
+    ASSERT(client_map_.empty());
+    ASSERT(host_set_member_update_cb_handle_ == nullptr);
+    return nullptr;
+  }
+
+  auto colon_pos = host_address.rfind(":");
+  if ((colon_pos == std::string::npos) || (colon_pos == (host_address.size() - 1))) {
+    return nullptr;
+  }
+
+  std::string ip_address = host_address.substr(0, colon_pos);
+  bool ipv6 = (ip_address.find(":") != std::string::npos);
+  std::string host_address_map_key;
+  Network::Address::InstanceConstSharedPtr address_ptr;
+
+  if (!ipv6) {
+    host_address_map_key = host_address;
+  } else {
+    // find our canonical IPv6 address string representation by
+    // creating the address from the Redis server info...
+    std::string ip_port = host_address.substr(colon_pos + 1);
+    uint64_t ip_port_number;
+    if (!StringUtil::atoull(ip_port.c_str(), ip_port_number) || (ip_port_number > 65535)) {
+      // bad port number conversion or out of range for a TCP port
+      return nullptr;
+    }
+    try {
+      address_ptr = std::make_shared<Network::Address::Ipv6Instance>(ip_address, ip_port_number);
+    } catch (const EnvoyException&) {
+      // bad IPv6 address
+      return nullptr;
+    }
+    host_address_map_key = address_ptr->asString();
+  }
+
+  auto it = host_address_map_.find(host_address_map_key);
+  if (it == host_address_map_.end()) {
+    // not known to the cluster manager, create a new host and insert it into the map
+    if (!ipv6) {
+      // only create an IPv4 address instance if we need a new Upstream::HostImpl
+      std::string ip_port = host_address.substr(colon_pos + 1);
+      uint64_t ip_port_number;
+      if (!StringUtil::atoull(ip_port.c_str(), ip_port_number) || (ip_port_number > 65535)) {
+        // bad port number conversion or out of range for a TCP port
+        return nullptr;
+      }
+      try {
+        address_ptr = std::make_shared<Network::Address::Ipv4Instance>(ip_address, ip_port_number);
+      } catch (const EnvoyException&) {
+        // bad IPv4 address
+        return nullptr;
+      }
+    }
+    Upstream::HostSharedPtr new_host{new Upstream::HostImpl(
+        cluster_->info(), "", address_ptr, envoy::api::v2::core::Metadata::default_instance(), 1,
+        envoy::api::v2::core::Locality(),
+        envoy::api::v2::endpoint::Endpoint::HealthCheckConfig::default_instance(), 0,
+        envoy::api::v2::core::HealthStatus::UNKNOWN)};
+    host_address_map_[host_address_map_key] = new_host;
+    it = host_address_map_.find(host_address_map_key);
+  }
+
+  ThreadLocalActiveClientPtr& client = client_map_[it->second];
```

Comment:
An upstream redis can return an IP that we don't know about in our cluster, right? If that's the case a misbehaving upstream could give us an unlimited number of IPs that we never purge? Can you talk through the misbehaving upstream scenario and whether we need to protect against it in some way? Also, how do IPs/connections that aren't in our cluster work out?

Reply:
You are correct. The Redis server(s) we know about could redirect us to other servers that may or may not be in the proxy's configuration. This is not the same as random downstream connections to the proxy, however; the upstream server(s) in the configuration are at least known, if not fully trusted. Being redirected to IP addresses and TCP ports that are not being serviced means we might have a large number of embryonic TCP connections (equal to the number of unique IP address and TCP port tuples). As far as problems with the redis cluster are concerned, it is much more likely that we'll be redirected to a small set of non-operable address/port combinations many times. In that case, the requests will queue, and eventually error out upon connection timeout.

When @HenryYYang implements "cluster slots" cluster updates (which he's working on), we can then sensibly scan upstream connections to unknown hosts (i.e., not reported in the cluster slots output), report and log them, and then close and cancel any associated requests.

Reply:
I think ultimately we need to track how many "detached" connections we have via stats, and probably eventually make sure we cap the number of connections, time them out and clean them, etc. This PR is being enough so probably not worth doing that here, but can you add relevant TODOs that we can track for future work?

Reply:
adding // TODO